internal/pkg/addon/output.go

```diff
@@ -25,52 +29,127 @@ type Output struct {
-			"Ref": func(_ string, input interface{}, _ interface{}) interface{} {
-				if logicalID, ok := input.(string); ok {
-					return logicalID
-				}
-				return nil
-			},
-		},
-	})
+	type cfnTemplate struct {
+		Resources yaml.Node `yaml:"Resources"`
+		Outputs   yaml.Node `yaml:"Outputs"`
+	}
+	var tpl cfnTemplate
+	if err := yaml.Unmarshal([]byte(template), &tpl); err != nil {
+		return nil, fmt.Errorf("unmarshal addon cloudformation template: %w", err)
+	}
+
+	typeFor, err := parseTypeByLogicalID(&tpl.Resources)
+	if err != nil {
+		return nil, err
+	}
+
+	outputNodes, err := parseOutputNodes(&tpl.Outputs)
 	if err != nil {
-		return nil, fmt.Errorf("parse CloudFormation template %s: %w", template, err)
+		return nil, err
 	}
 
 	var outputs []Output
-	for logicalID, output := range tpl.Outputs {
-		outputs = append(outputs, Output{
-			Name:            logicalID,
-			IsSecret:        isSecret(output, tpl.GetAllSecretsManagerSecretResources()),
-			IsManagedPolicy: isManagedPolicy(output, tpl.GetAllIAMManagedPolicyResources()),
-		})
+	for _, outputNode := range outputNodes {
+		output := Output{
+			Name:            outputNode.name(),
+			IsSecret:        false,
+			IsManagedPolicy: false,
+		}
+		ref, ok := outputNode.ref()
+		if ok {
+			output.IsSecret = typeFor[ref] == secretManagerSecretType
+			output.IsManagedPolicy = typeFor[ref] == iamManagedPolicyType
+		}
+		outputs = append(outputs, output)
 	}
 	return outputs, nil
 }
 
-func isSecret(output cloudformation.Output, secrets map[string]*secretsmanager.Secret) bool {
-	value, ok := output.Value.(string)
-	if !ok {
-		return false
+// parseTypeByLogicalID returns a map where the key is the resource's logical ID and the value is the CloudFormation Type
+// of the resource such as "AWS::IAM::Role".
+func parseTypeByLogicalID(resourcesNode *yaml.Node) (typeFor map[string]string, err error) {
+	if resourcesNode.Kind != yaml.MappingNode {
+		// "Resources" is a required field in CloudFormation, check if it's defined as a map.
+		return nil, errors.New(`"Resources" field in cloudformation template is not a map`)
+	}
+
+	// The content of a !!map, like the "Resources" field, always come in pairs.
+	// The first element represents the key, ex: {Value: "Cluster", Kind: ScalarNode, Tag: "!!str", Content: nil}
+	// The second element holds the values such as "Type" and "Properties", ex: {Value: "", Kind: MappingNode, Tag:"!!map", Content:[...]}
+	typeFor = make(map[string]string)
+	for i := 0; i < len(resourcesNode.Content); i += 2 {
+		logicalIDNode := resourcesNode.Content[i]
+		fieldsNode := resourcesNode.Content[i+1]
+		fields := struct {
+			Type string `yaml:"Type"`
+		}{}
+		if err := fieldsNode.Decode(&fields); err != nil {
+			return nil, fmt.Errorf(`decode the "Type" field of resource "%s": %w`, logicalIDNode.Value, err)
+		}
+		typeFor[logicalIDNode.Value] = fields.Type
+	}
+	return typeFor, nil
+}
+
+func parseOutputNodes(outputsNode *yaml.Node) ([]*outputNode, error) {
```

Comment:
same for this function

Reply:
OK done for both! https://github.com/aws/copilot-cli/pull/1804/commits/7a864d46fd61c3cba28b73132d400c5ef6fdac53